src/main/scala/org/apache/spark/sql/delta/storage/GcsLogStore.scala

```diff
@@ -0,0 +1,92 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.delta.storage
+
+import java.io.{IOException, _}
+import java.nio.charset.StandardCharsets.UTF_8
+import java.nio.file.FileAlreadyExistsException
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.hadoop.fs._
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+
+/**
+ * The [[LogStore]] implementation for Gcs, which uses gcs-connector to
+ * provide the necessary atomic and durability guarantees:
+ *
+ * 1. Atomic Visibility: Read/read-after-metadata-update/delete are strongly
+ * consistent for gcs.
+ *
+ * 2. Consistent Listing: Gcs guarantees strong consistency for both object and
+ * bucket listing operations.
+ * https://cloud.google.com/storage/docs/consistency
+ *
+ * 3. Mutual Exclusion: Preconditions are used to handle race conditions.
+ *
+ * Regarding file creation, this implementation:
+ * - Throws [[FileAlreadyExistsException]] if file exists and overwrite is false.
+ * - Opens a stream to write to gcs otherwise.
+ * - Assumes file writing to be all-or-nothing, irrespective of overwrite option.
+ */
+class GcsLogStore(sparkConf: SparkConf, defaultHadoopConf: Configuration)
+  extends HadoopFileSystemLogStore(sparkConf, defaultHadoopConf) with Logging {
+
+  protected def getFileContext(path: Path): FileContext = {
+    FileContext.getFileContext(path.toUri, getHadoopConfiguration)
+  }
+
+  val preconditionFailedExceptionMessage = "412 Precondition Failed"
+
+  def write(path: Path, actions: Iterator[String], overwrite: Boolean = false): Unit = {
+    val fs = path.getFileSystem(getHadoopConfiguration)
+    if (!fs.exists(path.getParent)) {
+      throw new FileNotFoundException(s"No such file or directory: ${path.getParent}")
+    }
+
+    if (fs.exists(path) && !overwrite) {
+      throw new FileAlreadyExistsException(path.toString)
+    }
+    writeObject(path, fs, actions)
+  }
+
+  private def writeObject(path: Path, fs: FileSystem, actions: Iterator[String]): Unit = {
+    var streamClosed = false // This flag is to avoid double close
+    val stream = fs.create(path, true)
+    try {
+      actions.map(_ + "\n").map(_.getBytes(UTF_8)).foreach(stream.write)
+      stream.close()
+      streamClosed = true
+    } catch {
+      // Gcs uses preconditions that guarantee that object will be created iff it doesn't exist.
+      // Reference: https://cloud.google.com/storage/docs/generations-preconditions
+      case e: IOException if e.getCause.getMessage.contains(preconditionFailedExceptionMessage) =>
+        val newException = new FileAlreadyExistsException(path.toString, null, e.getMessage)
```

Comment:
Does Delta require that we throw `FileAlreadyExistsException` here? It could be wrong to throw this exception here if we expect to overwrite existing file.

Reply:
Delta Lake java doc says that `FileAlreadyExistsException` should be thrown only if `overwrite` is `false`: https://github.com/delta-io/delta/blob/3f518a2d56784de362151a115e920e2e4050b59d/src/main/scala/org/apache/spark/sql/delta/storage/LogStore.scala#L97:L102

Reply:
should we add a retry logic in this scenario?

Reply:
1) stream = fs.create
2) stream.close
File will only be visible after 2.
 
If file exists before 1 and overwrite=false then method will throw java.nio.file.FileAlreadyExistsException and will successfully write if overwrite=true. This basically handles atomicity of write through all-or-nothing.

Now if file gets created between 1 and 2, then irrespective of overwrite flag, pre-condition failure will happen. This can happen due to multiple writers. DeltaLake expects mutual exclusion for writers. So we might need to come up with an exception free approach for this scenario.

In case of other logstores:
i) HDFSLogStore: FileContext.rename is used to move temporary file(with contents) to requested path. Doesn't have any specific logic for race conditions. Expects only org.apache.hadoop.fs.FileAlreadyExistsException which gets converted to java.nio.file.FileAlreadyExistsException

ii) HadoopFileSystemLogStore/AzureLogStore: Uses FileSystem.rename and rest is similar as above but if rename fails but fs.exists(path) is false then throws IllegalStateException.

iii) S3SingleDriverLogStore: Uses lock on paths and doesn't have specific logic for race conditions.

Reply:
I think that to be compatible with Delta Lake contract we should throw exception if overwrite is false regardless of when file was created before `1` or between `1` and `2`, in all other cases we should ignore precondition exception.

Reply:
will handle overwrite=false by throwing FileAlreadyExistsException but i don't understand what do you mean by ignore, suppress or throw as it is.

Also, wouldn't allowing file creation between 1 and 2 will violate the contract for "Mutual exclusion of writers" required by deltalake from file system.

Reply:
I meant to suppress it.

Regarding "Mutual exclusion", from javadoc it's unclear what does it mean for overwrite case? Does Delta Lake any docs/tests that describe how it should behave in case of overwrite and concurrent modification?

Reply:
Changed to suppression for `overwrite=true`. 

I couldn't find any documentation/javadoc for error-handling in case of `overwrite=true`. Also tests are pretty straight-forward and doesn't go into much details. Currently for `overwrite=false` `GcsLogStore` will throw `FileAlreadyExistsException` which is handled by [transaction commit logic][1] by trying iteratively with increasing `attemptNumber`. This takes care into the scenario where write-with-append by simply attempting again whereas for write-with-overwrite this throws `ConcurrentTransactionException`.

As we are suppressing exception for `overwrite=true`, any written data for this transaction will be lost(subsequent reads can not access it) as no log store commits are made. In delta-lake commits with `overwrite=true` is present for [checkpoint creation][2] but I couldn't find any error-handling logic there.
 

[1]: https://github.com/delta-io/delta/blob/master/src/main/scala/org/apache/spark/sql/delta/OptimisticTransaction.scala#L562-L584
[2]: https://github.com/delta-io/delta/blob/master/src/main/scala/org/apache/spark/sql/delta/Checkpoints.scala#L127